### DISCUSSION TITLE
how to implement refresh token?

### DISCUSSION THREAD

**Discussion by ShourovRoy** [zino-app/graphql-flutter#1066] [Q&A] [answered]
how to implement refresh token?

**Comment by vincenzopalazzo** [comment]
Hi, sorry for the delay here.

Can you make a small example of the use case? for now, my suggestion is to create a new client to refresh the token stored inside it or just update the link inside the client.

**Reply by ShourovRoy** [reply]
I was looking to use refresh token to get new access token but didn't find any official docs with flutter graphql.

**Reply by vincenzopalazzo** [reply]
This PR can help with your feature request? https://github.com/zino-hofmann/graphql-flutter/pull/1081

basically with this PR, you could do stuff like 

```
    var link = HttpLink(endPoint, defaultHeaders: {
      'Authorization': 'bearer $token',
    });
    client.updateClient(link: link);
```

**Reply by ShourovRoy** [reply]
This is not what I want. Actually, is there any interceptor in flutter_graphql like javascript axios has?? I can check the response if the access token expired then I can re-request with refresh token to get new access token? Is that possible  in flutter_graphql?

**Reply by budde377** [reply]
I don't think that is something the client should handle. Instead, you should implement a link with your desired refresh logic.

**Reply by Kiruel** [reply]
Do we have an example of that ? I try using http_interceptor but never pass the the `interceptRequest`:
```
GraphQLClient(
        cache: GraphQLCache(),
        link: HttpLink(
          MY_UrL,
          httpClient: InterceptedClient.build(
            interceptors: [
              LoggingInterceptor(
                  authLocalDataSource: authLocalDataSource,
              ),
            ],
            client: http.Client(),
          ),
        ),
      )
 ```
The interceptor:
```
class LoggingInterceptor implements InterceptorContract {
  LoggingInterceptor({
    required this.authLocalDataSource,
  });

  final AuthLocalDataSource authLocalDataSource;

  @override
  Future<RequestData> interceptRequest({required RequestData data}) async {
    try {
      final result = await authLocalDataSource.getCredentials();
      data.headers['Authorization'] = 'Bearer ${result.accessToken}';
    } catch (e) {
      return data;
    }
    return data;
  }

  @override
  Future<ResponseData> interceptResponse({required ResponseData data}) async {
    return data;
  }
}
```
 I'm looking for a simple example of that. Maybe I make it wrong.

**Reply by Kiruel** [reply]
Finaly I find a way to do it with gql_dio_link :
```
GraphQLClient(
        cache: GraphQLCache(),
        link: DioLink(
          'graphql',
          client: dio,
        ),
      )
```
And the dio :
```
Dio(
        BaseOptions(
          baseUrl: URL_SERVER,
        ),
      )..interceptors.add(OauthInterceptor(
          authLocalDataSource: authLocalDataSource,
        ));
```

**Reply by vincenzopalazzo** [reply]
@ShourovRoy The code is meaningless here, why do you will not a good format PR to improve the docs? we love this type of PR

**Reply by Kiruel** [reply]
I will make a PR for the doc when I got some time ;) !

**Comment by vincenzopalazzo** [comment, marked as answer]
The solution was reported by this answer https://github.com/zino-hofmann/graphql-flutter/discussions/1066#discussioncomment-3066109
